pkg/controllers/clusterresourceplacement/placement_status.go

```diff
@@ -0,0 +1,319 @@
+		}
+
+		workKObj := klog.KObj(&workList.Items[i])
+		indexFromWork, err := labels.ExtractResourceSnapshotIndexFromWork(&workList.Items[i])
+		if err != nil {
+			klog.ErrorS(err, "Failed to parse the resource snapshot index label from work", "clusterResourcePlacement", crpKObj, "work", workKObj)
+			return controller.NewUnexpectedBehaviorError(err)
+		}
+		if indexFromWork > resourceIndex {
+			err := fmt.Errorf("invalid work %s: resource snapshot index %d on the work is greater than resource index %d on the latest clusterResourceSnapshot", workKObj, indexFromWork, resourceIndex)
+			klog.ErrorS(err, "Invalid work", "clusterResourcePlacement", crpKObj, "clusterResourceSnapshot", klog.KObj(latestResourceSnapshot), "work", workKObj)
+			return controller.NewUnexpectedBehaviorError(err)
+		} else if indexFromWork < resourceIndex {
+			// The work is pointing to the old resourceSnapshot.
+			// it means the rollout controller has not updated the binding yet or work generator has not handled this work yet.
+			oldWorkCounter++
+		} else { // indexFromWork = resourceIndex
+			// The work is pointing to the latest resourceSnapshot.
+			newWorkCounter++
+		}
+		// we build the applied condition on all the works no matter they're latest or not.
+		isPending, failedManifests := buildFailedResourcePlacements(&workList.Items[i])
+		if isPending {
+			pendingWorkCounter++
+		}
+		if len(failedManifests) != 0 && len(failedResourcePlacements) <= maxFailedResourcePlacementLimit {
+			failedResourcePlacements = append(failedResourcePlacements, failedManifests...)
+		}
+	}
+
+	if len(failedResourcePlacements) > maxFailedResourcePlacementLimit {
+		failedResourcePlacements = failedResourcePlacements[0:maxFailedResourcePlacementLimit]
+	}
+
+	klog.V(2).InfoS("Building the resourcePlacementStatus", "clusterResourcePlacement", crpKObj, "clusterName", status.ClusterName,
+		"numberOfOldWorks", oldWorkCounter, "numberOfNewWorks", newWorkCounter,
+		"numberOfPendingWorks", pendingWorkCounter, "numberOfFailedResources", len(failedResourcePlacements))
+
+	status.FailedResourcePlacements = failedResourcePlacements
+
+	desiredWorkCounter, err := annotations.ExtractNumberOfResourceSnapshotsFromResourceSnapshot(latestResourceSnapshot)
+	if err != nil {
+		klog.ErrorS(err, "Master resource snapshot has invalid numberOfResourceSnapshots annotation", "clusterResourcePlacement", crpKObj, "clusterResourceSnapshot", klog.KObj(latestResourceSnapshot))
+		return controller.NewUnexpectedBehaviorError(err)
+	}
+	workSynchronizedCondition, err := buildWorkSynchronizedCondition(crp, desiredWorkCounter, newWorkCounter, oldWorkCounter)
+	if err != nil {
+		return err
+	}
+	meta.SetStatusCondition(&status.Conditions, workSynchronizedCondition)
+	if newWorkCounter+oldWorkCounter == 0 {
+		// skip setting the applied work condition
+		return nil
+	}
+	workAppliedCondition := buildWorkAppliedCondition(crp, pendingWorkCounter > 0, len(failedResourcePlacements) > 0)
```

Comment:
can  newNworkCount < desiredWorkCount but pendingWorkCount == 0 and no failedPlacement? It seems that if the sync condition is not true then we can always set appliedCondition as "unknown" or "false" depends on you interpret "unknown"

Reply:
can newNworkCount < desiredWorkCount but pendingWorkCount == 0 and no failedPlacement?

it could, eg, desiredWorkCount = 5, we successfully create 3 new works.

I updated the status implementation here and build the applied status only on the newWork.
The appliedCondition will be unknown if the syncCondition is not true.